### PR TITLE
Emily/def

### PR DESCRIPTION
**Pull request by egiurleo**
### Test plan

See included automated tests.

### CODE REVIEW

**Review by amomchilov — COMMENTED** [review]
The existing code LGTM, but I thought of some other edge cases to consider:

```rb
def anon_rest(*); end
def anon_kwarg(**); end
def anon_bloc(&); end
def forward_everything(...); end
def prohibit_kwargs(**nil); end
def prohibit_block(&nil); end
```

File: main/pipeline/pipeline.cc [hunk 1]
```diff
@@ -203,6 +239,82 @@ unique_ptr<parser::Node> convertPrismToSorbet(pm_node_t *node, pm_parser_t *pars
...
+        }
+        case PM_PARAMETERS_NODE: {
+            auto paramsNode = reinterpret_cast<pm_parameters_node *>(node);
+            pm_location_t *loc = &paramsNode->base.location;
+
+            pm_node_list_t *requireds = &paramsNode->requireds;
+            pm_node_list_t *optionals = &paramsNode->optionals;
+            pm_node_list_t *keywords = &paramsNode->keywords;
+
+            parser::NodeVec params;
```
[the remark this answers is not in the document]
Comment (egiurleo) [reply main/pipeline/pipeline.cc:280]: Oh excellent, thanks for pointing this out.

**Comment by egiurleo** [comment]
@amomchilov I used your approach of iterating with `absl::span`s, and then I tried to clean it up even further by adding all params to a vector, converting that vector to a span and then iterating over that? I'm not sure if that last refactor actually helps though. Let me know what you think?

**Review by amomchilov — COMMENTED** [review]
Making the `std::vector<pm_node *>` really cleans up the code (with those relatively short/clean `allParams.insert` statements), but it introduces at least one allocation (perhaps several, because of the vector growth)

What you had in https://github.com/Shopify/sorbet/pull/33/commits/fcd8532064ff5b6c0e96231c1b9d759face353f8 is a bit more verbose, but skips this allocation(s) 

What's our stance on the the code verbosity vs runtime perf trade-off? Ofc we can/should measure this at some point, but given how common method defs are, I'd suspect that we'd want to avoid the allocations there

File: main/pipeline/pipeline.cc [hunk 2]
```diff
@@ -231,21 +231,37 @@ unique_ptr<parser::Node> convertPrismToSorbet(pm_node_t *node, pm_parser_t *pars
...
+                                               locOffset(nameLoc, parser), std::move(value));
+        }
         case PM_PARAMETERS_NODE: {
             auto paramsNode = reinterpret_cast<pm_parameters_node *>(node);
             pm_location_t *loc = &paramsNode->base.location;
 
-            pm_node_list_t *requiredsList = &paramsNode->requireds;
-            auto requireds = absl::MakeSpan(requiredsList->nodes, requiredsList->size);
+            auto requireds = absl::MakeSpan((&paramsNode->requireds)->nodes, (&paramsNode->requireds)->size);
+            auto optionals = absl::MakeSpan((&paramsNode->optionals)->nodes, (&paramsNode->optionals)->size);
```
Comment (amomchilov) [inline main/pipeline/pipeline.cc]: `(&a->b)->c` can be simplified to just `a->b.c`

```suggestion
            auto requireds = absl::MakeSpan(paramsNode->requireds.nodes, paramsNode->requireds.size);
            auto optionals = absl::MakeSpan(paramsNode->optionals.nodes, paramsNode->optionals.size);
```

Derivation:

```c
(&a->b)->c`
(&(a->b))->c` // 1. Add `()` to clarify binding of `&`
(*&(a->b)).c` // 2. `x->y` is shorthand for `(*x).y`
(a->b).c      // 3. `*&x` cancels out to x
              //     "take the address of x, then find what's pointed to by it"
              //      ... it's just x.
a->b.c        // 4. remove unnecessary `()`
```

**Review by amomchilov — APPROVED** [review]
Both variations being compared look good to me. We can always tweak stuff after.

I'll be away on Monday, so feel free to ship without me, so you don't have to stack on top of this PR

**Comment by egiurleo** [comment]
@amomchilov Mm yes I like your point about the allocations vs. verbosity tradeoff. I think we should optimize for performance and if the code really gets out of hand we can reevaluate. I'm going to revert to the previous version.
